src/jit/compiler.hpp

```diff
@@ -4711,6 +4711,29 @@ inline void Compiler::CLR_API_Leave(API_ICorJitInfo_Names ename)
 
 #endif // MEASURE_CLRAPI_CALLS
 
+//------------------------------------------------------------------------------
+// fgStructTempNeedsExplicitZeroInit : Check whether struct of the given type needs
+//                                     explicit zero initialization in this basic block.
+//
+// Arguments:
+//    structType -       type of the struct
+//    block  -           basic block to check
+//
+// Returns:
+//             true if a struct temp of the given type needs explicit zero-initialization in this basic block;
+//             false otherwise
+//
+// Notes:
+//     Structs with GC pointer fields are fully zero-initialized in the prolog if compInitMem is true.
+//     Therefore, we don't need to insert zero-initialization if this block is not in a loop.
+
+bool Compiler::fgStructTempNeedsExplicitZeroInit(CORINFO_CLASS_HANDLE structType, BasicBlock* block)
+{
+    DWORD typeFlags     = info.compCompHnd->getClassAttribs(structType);
+    bool  containsGCPtr = ((typeFlags & CORINFO_FLG_CONTAINS_GC_PTR) != 0);
```

Comment:
Can you verify this is the right test? See for instance `impNormStructType` where we also look at `CORINFO_FLG_CONTAINS_STACK_PTR`.

Reply:
Yes, this was missing some structs. I modified the check to match what's used in codegen to decide whether to zero-initialize the struct in the prolog. The codesize improvement is slightly better now.

Reply:
Frameworks diffs:
```
Summary:
(Note: Lower is better)
Total bytes of diff: -16580 (-0.13 % of base)
    diff is an improvement.
Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes
Top file improvements by size (bytes):
       -4874 : Microsoft.CodeAnalysis.CSharp.dasm (-0.22 % of base)
       -4566 : Microsoft.CodeAnalysis.dasm (-0.58 % of base)
       -2109 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.09 % of base)
       -1705 : System.Collections.Immutable.dasm (-1.18 % of base)
       -1119 : System.Reflection.Metadata.dasm (-1.45 % of base)
27 total files with size differences (27 improved, 0 regressed), 52 unchanged.
Top method regessions by size (bytes):
          23 : Microsoft.CodeAnalysis.VisualBasic.dasm - Conversions:ClassifyQueryLambdaConversion(ref,ref,ref,byref):int
Top method improvements by size (bytes):
       -2957 : Microsoft.CodeAnalysis.dasm - AttributeDescription:.cctor()
       -1094 : System.Reflection.Metadata.dasm - MetadataReader:InitializeProjectedTypes()
        -334 : System.Collections.Immutable.dasm - Enumerator:MoveNext():bool:this (19 methods)
        -312 : Microsoft.CodeAnalysis.CSharp.dasm - OverloadResolution:GetEnumOperation(int,ref,ref,ref,ref):this
        -308 : Microsoft.CodeAnalysis.dasm - <GetVerStrings>d__22:MoveNext():bool:this
452 total methods with size differences (451 improved, 1 regressed), 66047 unchanged.
```